src/fluree/db/query/optimize.cljc

```diff
@@ -36,3 +44,184 @@
+        highly-selective
+        moderately-selective)
+
+      :else
+      (let [[s p o] pattern-data]
+        (cond
+          ;; Class patterns use class count from stats
+          (= :class pattern-type)
+          (let [class-iri (get-iri o)
+                class-sid (encode-iri-to-sid db class-iri)]
+            (or (get-class-count stats class-sid) default-selectivity))
+
+          ;; Specific s-p-o triple lookup
+          (all-matched? [s p o])
+          highly-selective
+
+          ;; s-p-? lookup uses property count
+          (and (matched? s) (matched? p) (unmatched? o))
+          (let [pred-iri (get-iri p)
+                pred-sid (encode-iri-to-sid db pred-iri)]
+            (or (get-property-count stats pred-sid) moderately-selective))
+
+          ;; ?-p-? property scan uses property count
+          (and (unmatched? s) (matched? p) (unmatched? o))
+          (let [pred-iri (get-iri p)
+                pred-sid (encode-iri-to-sid db pred-iri)]
+            (or (get-property-count stats pred-sid) default-selectivity))
+
+          ;; ?-p-o reverse lookup (find subjects with specific value)
+          (and (unmatched? s) (matched? p) (matched? o))
+          highly-selective
+
+          (and (unmatched? s) (unmatched? p) (unmatched? o))
+          full-scan
+
+          :else
+          default-selectivity)))))
+
+(defn split-by-optimization-boundaries
+  "Split where clause into segments separated by optimization boundaries.
+   Optimizable patterns (:tuple, :class, :id) are grouped together.
+   All other patterns (boundaries) separate the segments."
+  [where-clause]
+  (reduce
+   (fn [segments pattern]
+     (if (optimizable-pattern? pattern)
+       (if (and (seq segments)
+                (= :optimizable (:type (peek segments))))
+         (update segments (dec (count segments))
+                 update :data conj pattern)
+         (conj segments {:type :optimizable :data [pattern]}))
+       ;; Boundary - add as separate segment
+       (conj segments {:type :boundary :data pattern})))
+   []
+   where-clause))
+
+(defn optimize-segment
+  "Optimize a single segment by sorting patterns by selectivity"
+  [db stats patterns]
+  (if (or (nil? stats) (empty? stats))
+    ;; No stats, return as-is
+    patterns
+    ;; Sort by selectivity (lower = more selective = execute first)
+    (let [with-scores (mapv (fn [pattern]
+                              {:pattern pattern
+                               :score (or (calculate-selectivity db stats pattern)
+                                          default-selectivity)})
+                            patterns)
+          sorted      (sort-by :score with-scores)]
+      (mapv :pattern sorted))))
+
+(defn optimize-patterns
+  "Reorder patterns for optimal execution based on statistics.
+   Splits on optimization boundaries and optimizes each segment independently."
+  [db where-clause]
+  (let [stats (:stats db)]
+    (if (or (nil? stats) (empty? stats))
```

Comment:
I don't think we should need to keep checking that we actually have `stats`, that should be handled once upstream.

Reply:
addressed in 3abcd39a4d59d0cbf51ac0d3b1a3e7bf6cdfe5d5